cuda/base/math.hpp

```diff
@@ -93,17 +94,126 @@ __device__ GKO_INLINE std::complex<double> one<std::complex<double>>()
 }
 
 
-// Since the `isfinite` function from CUDA is not in any namespace, it
-// will be used without the `using` keyword.
+// This first part is specific for clang in combination with the nvcc compiler
+// from the toolkit older than 9.2.
+// clang wants to use their `__builtin_isfinite` function, which is not present
+// as a __device__ function, so it results in a compiler error.
+// Here, `isfinite` is written by hand, which might not be as performant as the
+// intrinsic function from CUDA, but it compiles and works.
+#if defined(__CUDA_ARCH__) && defined(__CUDACC_VER_MAJOR__) &&     \
+    defined(__CUDACC_VER_MINOR__) &&                               \
+    (__CUDACC_VER_MAJOR__ * 1000 + __CUDACC_VER_MINOR__) < 9002 && \
```

Comment:
Is there any reason to use the combination of `__CUDACC_VER_MAJOR__` and `__CUDACC_VER_MINOR__`?
I prefer to use `CUDA_VERSION` (e.g. 9020 for 9.2, 10000 for 10.0).
`CUDA_VERSION` is defined in `cuda.h` but I can not find `__CUDACC_VER_MAJOR__` and `__CUDACC_VER_MINOR__`.

Reply:
Now, I can use `CUDA_VERSION`, but previously, I did not want to include `cuda.h` inside the public `math.hpp` header.
The `CUDACC_VER_`... are defined by the NVCC compiler ([documented here](https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html#nvcc-identification-macro)). At the time (and even now), I also did not find an official documentation of `CUDA_VERSION`, which is why I also left the current definition in.
Do you know where it is officially documented?

Reply:
That's the only documentation I can find.
https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__TYPES.html#group__CUDA__TYPES_1g3c09bba9b1547aa69f1e346b82bcdb50

Reply:
Thanks!
It is weird that they do not state that anywhere else in their documentation. I found `CUDA_VERSION` once in the programming guidelines, but not explained what the content is.

According to the site you linked, `CUDA_VERSION` is the
> CUDA API version number

Since I need the version of the actual compiler, and not the API version, I think it is better to leave the current Macros, since they are tied to the compiler.

Reply:
That's fine with me.

Reply:
It is okay to me